lib/Conversion/TritonGPUToLLVM/GatherOpToLLVM.cpp

```diff
@@ -0,0 +1,112 @@
+#include "triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
+
+using namespace mlir;
+using namespace mlir::triton;
+
+namespace {
+class GatherOpConversion : public ConvertOpToLLVMPattern<GatherOp> {
+public:
+  GatherOpConversion(LLVMTypeConverter &typeConverter,
+                     const TargetInfoBase &targetInfo, PatternBenefit benefit)
+      : ConvertOpToLLVMPattern(typeConverter, benefit), targetInfo(targetInfo) {
+  }
+
+  LogicalResult
+  matchAndRewrite(GatherOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override;
+
+private:
+  const TargetInfoBase &targetInfo;
+};
+
+LogicalResult
+GatherOpConversion::matchAndRewrite(GatherOp op, OpAdaptor adaptor,
+                                    ConversionPatternRewriter &rewriter) const {
+  Location loc = op.getLoc();
+  RankedTensorType srcType = op.getSrc().getType();
+
+  // Compute the src subtensor shape owned by this CTA.
+  SmallVector<unsigned> srcShapePerCTA =
+      convertType<unsigned>(triton::gpu::getShapePerCTA(srcType));
+
+  // Grab the src values in this thread.
+  SmallVector<Value> srcValues =
+      unpackLLElements(loc, adaptor.getSrc(), rewriter);
+
+  // Emit the indices of the src values owned by this thread.
+  SmallVector<SmallVector<Value>> srcIndices =
+      emitIndices(loc, rewriter, targetInfo, srcType.getEncoding(),
+                  op.getSrc().getType(), /*withCTAOffset=*/true);
+
+  // Store the src values owned by the thread into their respective location in
+  // the scratch memory.
+  assert(srcValues.size() == srcIndices.size());
+
+  // Get the base pointer to the scratch memory.
+  Value smemBase = LLVM::getSharedMemoryBase(loc, rewriter, targetInfo, op);
+
+  // For each src element owned by the thread, index into the scratch memory and
+  // then store it.
+  Type elemType = getTypeConverter()->convertType(srcType.getElementType());
+  for (auto [value, indices] : llvm::zip(srcValues, srcIndices)) {
+    // Convert the index at each dim into a single offset given the shape of the
+    // tensor.
+    Value offset = LLVM::linearize(rewriter, loc, indices, srcShapePerCTA);
+    // Emit the offset into the shared memory and then store the value.
+    Value ptr = gep(smemBase.getType(), elemType, smemBase, offset);
+    store(value, ptr);
```

Comment:
Can be left as a TODO, but we should do a masked store with `getRedundantDataMask`
https://github.com/triton-lang/triton/blob/dbebe10216c2dd546d3fdb3c8222939c94c8f8ce/third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp#L32

(TBH I think there are quite a few places we need to cleanup redundant operations)

Reply:
I don't think a masked store is better, in general there isn't performance penalty for storing multiple times to the same address in shared memory. Using a store without mask allows better code generation in general

Reply:
>  in general there isn't performance penalty for storing multiple times to the same address in shared memory.

would I be right to say this is only true if there are no bank conflicts?

Reply:
no, as far as I know it is orthogonal to bank conflicts

Reply:
I'm confused. Are you saying that  `@p st.shared` where `p` is false on all threads in the warp can still take multiple cycles?

Reply:
ah no that's not what I meant. I meant when doing a `st.shared` with duplicated addresses there isn't a penalty for storing multiple time. The HW will pick one of them and store only once.

Reply:
Yes I agree that on a single instruction level redundancy is okay. I guess that could be expressed by ignoring redundancy within a warp.

The problematic cases are:
- within a thread i.e. calling `st.shared` more times than necessary.
- between warps i.e. multiple warps calling `st.shared` to transfer the same data.

Reply:
true. Also another problem I remember running into with predicates (although not this exact case) is that it tends to throw off ptxas a lot in term of scheduling and liveranges. This applies more to `load` than to `store` (the liveness problem doesn't exist with loads, the scheduling I'm not sure) but that's one thing to keep in mind.